docs/proposals/scheduling/20231205-dynamic-reservation-provisioning.md

```diff
@@ -0,0 +1,215 @@
+
+Outcome with DRP: DRP facilitates on-demand scaling by provisioning reservations as microservices scale out. This ensures that each service instance has the resources it needs to perform optimally without pre-allocating resources.
+
+
+## Proposal
+
+### Overview
+
+The DRP mechanism will work as follows:
+
+1. Pod Annotation for Dynamic Reservation: Users can annotate Pods with `scheduling.koordinator.sh/reservation-template`, which includes the specification for a potential Reservation CRD.
+2. Scheduler Responsiveness: If an incoming Pod has this annotation and there is no available Reservation CRD instance that fits its requirements, the Koordinator scheduler will construct a virtual reservation object in memory. This object will emulate the behavior of a Reservation CRD instance to temporarily reserve resources for the Pod.
+3. Triggering Reservation Creation: In the PreBind phase of scheduling, the scheduler appends an annotation `scheduling.koordinator.sh/new-reservation-requirement` to the Pod, signaling an external provisioner to create an actual Reservation CRD instance based on the provided template.
+4. Reservation CRD Instance Provisioning: The external provisioner watches the Pod, once one Pod has annotation `scheduling.koordinator.sh/new-reservation-requirement` and `scheduling.koordinator.sh/reservation-template`, the provisioner will create a Reservation CRD instance with the annotation `scheduling.koordinator.sh/reservation-id` that defined in `new-reservation-requirement` to associate the virtual reservation object in scheduler. When the scheduler detects a newly created Reservation object with an associated virtual reservation object, it takes on the responsibility to update the status of the Reservation. It binds the Reservation to the node where the virtual reservation object was originally intended and marks the Reservation as available. 
+5. Completion of the Binding Process: After the scheduler receives notification of a new Reservation CRD instance, it replaces the virtual object with the actual reservation, adjusts the reservation allocable resources by an amount equal to the sum of the requests of the associated Pods, and continues to bind the Pod to the appropriate node.
+
+### Fault tolerance
+
+Once provisioner exception prevents reservation from being created, fault tolerance needs to be implemented:
+
+- The scheduler should have a default wait timeout, similar to the mechanism in Dynamic Volume Provisioning. If a Reservation is not created after waiting for 10 minutes, the process should be rolled back, and the Pod should be re-queued for another scheduling attempt.
+- Consider retaining the virtual reservation object; it could be reused in the next Pod scheduling attempt. It is recommended to keep it at first. Once the provisioner recovers from any exceptions, it can attempt to create a new Reservation object that can replace the virtual reservation object in memory.
```

Comment:
It seems unreasonable. If the Reservation is not created externally, this resource reserved by the object should not be used by the Pod at this time.